rs/utils/src/l2.rs

```diff
@@ -1,13 +1,100 @@
+use std::{
+    ops::Mul,
+    simd::{f32x4, f32x8, num::SimdFloat},
+};
+
 use crate::DistanceCalculator;
 
 pub struct L2DistanceCalculator {}
 
-impl DistanceCalculator for L2DistanceCalculator {
-    fn calculate(&self, a: &[f32], b: &[f32]) -> f32 {
+impl L2DistanceCalculator {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    #[allow(dead_code)]
+    fn calculate_non_simd(&self, a: &[f32], b: &[f32]) -> f32 {
         let mut dist = 0.0;
         for i in 0..a.len() {
             dist += (a[i] - b[i]).powi(2);
         }
         dist.sqrt()
     }
 }
+
+impl DistanceCalculator for L2DistanceCalculator {
+    fn calculate(&self, a: &[f32], b: &[f32]) -> f32 {
+        let mut i = 0;
+        let mut dist_8 = f32x8::splat(0.0);
+        let mut dist_4 = f32x4::splat(0.0);
+        let mut dist = 0.0;
+        while i + 8 <= a.len() {
```

Comment:
How about using `f32x8::lanes()` instead of hardcoding 8 here and below?
Same for `f32x4` too btw.

Reply:
hmm no such function

Reply:
sorry, should be `dist_8.len()`, not `lanes`